### PR TITLE
[CB] Add per-request logits processors

### PR DESCRIPTION
**Pull request by remi-or**
# Summary
This PR adds per-request logits processors and overalls the way CB handles logits processors. 
It introduces batched logits processing with per-request parameters for continuous batching, enabling each request in a batch to use different sampling parameters (temperature, top_k, top_p). This is essential for serving scenarios where different users may request different generation configurations within the same batch.

The main changes are:
  - cb_logits_processors.py (new): Adds `ContinuousBatchingLogitsProcessorList` to manage logits processors and three per-request processor implementations (Temperature, TopK, TopP) that operate on the batched tensor format using vectorized operations
  - continuous_api.py: Integrates the new processor list into the batch processor, passes it through the forward/sampling pipeline, and validates per-request kwargs at request ingestion
  - input_outputs.py: Extends the bulk input tensor to store processor arguments (one row per processor), passes them to PagedAttentionArgs, and prepares tensor args during batch preparation
  - requests.py: Adds logit_processor_kwargs field to RequestState for per-request overrides, reorganizes dataclass fields, and simplifies fork() using deepcopy
  - configuration_utils.py: Adds per_request_processors and drop_unsupported_processors flags to ContinuousBatchingConfig
  - logits_process.py: Adds supports_continuous_batching flag to base processors for compatibility checking
  - Tests: Extends test coverage for per-request temperature, top_k, and top_p sampling

  The processor list validates compatibility at init time, warns about unsupported processors, and efficiently prepares tensor arguments by storing them as views into the bulk input tensor to minimize memory transfers.

### CODE REVIEW

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+    TemperatureLogitsWarper,
+    TopKLogitsWarper,
+    TopPLogitsWarper,
+)
+from .requests import FutureRequestState, logger
+
+
+# Abstract base class for all continuous batching logits processors
+class ContinuousBatchingLogitsProcessor(ABC):
+    supported_kwargs: tuple[str, ...]  # Kwargs that this processor actively uses
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: should be a typedict

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+    if you want to use temperature scaling, you need to specify a temperature that's neither None nor 1.0. Otherwise
+    no processors will be created for temperature, and per-request temperature scaling will not be available.
+
+    On support of base processors:
+        Some base processors are not supported by CB and will be dropped when this class is instanciated. Some
+        processors have not yet been categorized as supported or not and will be kept but with a warning. All processors
+        can be kept by setting the flag `drop_unsupported_processors` to False.
+    On per-request processors:
+        Some base processors have a per-request version adapted for CB and will be converted to their per-request
+        version when this class is instanciated. Unless the flag `per_request_processors` is set to False.
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: why don't we default to True her?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+    """
+
+    def __init__(
+        self,
+        logits_processor: LogitsProcessorList,
+        per_request_processors: bool = False,
+        drop_unsupported_processors: bool = True,
+    ) -> None:
+        self.logits_processor = logits_processor
+        self.tensors_required = 0
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: Why do we count how many tensors ?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+        for i, processor in enumerate(self.logits_processor):
+            if isinstance(processor, TemperatureLogitsWarper):
+                self.logits_processor[i] = ContinuousBatchingTemperatureLogitsWarper(processor.temperature)
+                self.tensors_required += 1
+            elif isinstance(processor, TopKLogitsWarper):
+                self.logits_processor[i] = ContinuousBatchingTopKLogitsWarper(processor.top_k)
+                self.tensors_required += 1
+            elif isinstance(processor, TopPLogitsWarper):
+                self.logits_processor[i] = ContinuousBatchingTopPLogitsWarper(processor.top_p)
+                self.tensors_required += 1
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: these 3 are the only supported ones ATM right ? 
I think a mapping can help (keys().len(), registering etc?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        """
+        filtered_processors = []
+        for processor in self.logits_processor:
+            # Filter based on class: keep all ContinuousBatchingLogitsProcessor and error on non logits processors
+            class_name = processor.__class__.__name__
+            if isinstance(processor, ContinuousBatchingLogitsProcessor):
+                filtered_processors.append(processor)
+                continue
+            elif not isinstance(processor, LogitsProcessor):
+                raise TypeError(f"Processor {class_name} is not a logits processor.")
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: typing could be enough here no?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+                filtered_processors.append(processor)
+            # Case: processor is explicitly not supported (and maybe it is dropped)
+            else:
+                msg = f"Processor {class_name} is not supported by continuous batching."
+                if drop_unsupported:
+                    logger.warning(f"{msg} Dropping it.")
+                else:
+                    logger.warning(f"{msg} Keeping it because {drop_unsupported = }.")
+                    filtered_processors.append(processor)
+
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: looks kindaaaa bloated no?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+        if unknown_keys:
+            raise ValueError(
+                f"Unknown logit_processor_kwargs: {unknown_keys}. {self.supported_keys = } and {self.ignored_keys = }"
+                "If you expect a key to not be ignored, make sure its default value (in the generation config) is not "
+                "None. Eg. if temperature is None or 1.0 at creation time, no processor will be created for temperature"
+            )
+        # If there are none, throw a warning about the ignored keys
+        logger.warning(
+            f"Ignored logit_processor_kwargs: {problematic_keys}. {self.supported_keys = } and {self.ignored_keys = }"
+        )
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py:179]: w could just let it fail explicitly when an arg / kwargs is passed + good typing

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+
+    def prepare_tensor_args(
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py:189]: this is an important func!

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 9]
```diff
@@ -0,0 +1,305 @@
...
+                current_arg_id += 1
+            else:
+                scores = processor(input_ids, scores)
+        return scores
+
+
+# Here are all the continuous batching logits processors that are supported
+class ContinuousBatchingTemperatureLogitsWarper(ContinuousBatchingLogitsProcessor):
+    supported_kwargs: tuple[str, ...] = ("temperature",)
+    ignored_kwargs: tuple[str, ...] = ()
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: tbf i don't think we need all that

File: src/transformers/generation/continuous_batching/cb_logits_processors.py [hunk 10]
```diff
@@ -0,0 +1,305 @@
...
+class ContinuousBatchingTopPLogitsWarper(ContinuousBatchingLogitsProcessor):
+    supported_kwargs: tuple[str, ...] = ("top_p",)
+    ignored_kwargs: tuple[str, ...] = ("filter_value", "min_tokens_to_keep")
+
+    def __init__(self, top_p: float, filter_value: float = -float("Inf"), min_tokens_to_keep: int = 1):
+        top_p = float(top_p)
+        if top_p < 0 or top_p > 1.0:
+            raise ValueError(f"`top_p` has to be a float > 0 and < 1, but is {top_p}")
+        if not isinstance(min_tokens_to_keep, int) or (min_tokens_to_keep < 1):
+            raise ValueError(f"`min_tokens_to_keep` has to be a positive integer, but is {min_tokens_to_keep}")
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cb_logits_processors.py]: `@strict` decorator validation for this

File: src/transformers/generation/continuous_batching/input_outputs.py [hunk 11]
```diff
@@ -89,6 +93,8 @@ class is meant to be self-contained, so once a set of inputs have been created,
     batch alone.
     """
 
+    bulk_inputs: int = 7
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/input_outputs.py]: ? 7 ?

File: tests/generation/test_continuous_batching.py [hunk 12]
```diff
@@ -1028,3 +1051,133 @@ def test_flash_attn_with_kvcache_parity(self, use_cuda_graph: bool, use_async: b
...
+            tokenizer=tokenizer,
+            user_messages=user_messages,
+            logits_processor=logits_processor,
+            max_new_tokens=max_new_tokens,
+            do_sample=False,
+            output_scores=True,
+            eos_token_id=eos_token_id,
+        )
+        # Compare outputs
+        self.assertEqual(results[req1_id].generated_tokens, regular_generated_tokens[0])
```
Comment (ArthurZucker) [inline tests/generation/test_continuous_batching.py]: nice

File: src/transformers/generation/logits_process.py [hunk 13]
```diff
@@ -139,6 +137,8 @@ class MinLengthLogitsProcessor(LogitsProcessor):
     ```
     """
 
+    supports_continuous_batching: bool = False
```
Comment (ArthurZucker) [inline src/transformers/generation/logits_process.py:140]: should not be needed if we have a mapping / list (we don't "pollute" unrelated classes.

**Review by ArthurZucker — APPROVED** [review]
Nice! Missing examples in doc? but happy to have otherwise lets reduce bload in ContinuousBatchingLogitsProcessorList

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 1 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: Why? Do you want to add type checking when passing args?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 3 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: To know how much space to allocate for the per-request processors. I added a comment.

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 4 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: Agreed, added a mapping. 
This ` (keys().len(),` I did not understand sry

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 5 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: Yes, removed the error

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 6 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: Simplified it, only 4 cases now. Still pretty usefull to have comprehensive error tracking.

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 9 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: I think it's better if users of the API know which arguments are being used and which are not, and that we check for completely unsupported ones. Otherwise it's wild kwargs, no?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 10 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: `@strict` is for config I think? I dont see any logits processor with strict.

File: src/transformers/generation/continuous_batching/input_outputs.py (hunk 11 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/input_outputs.py]: Yup, changed name and added comment

File: tests/generation/test_continuous_batching.py (hunk 12 above)
Comment (remi-or) [reply tests/generation/test_continuous_batching.py]: thx

File: src/transformers/generation/logits_process.py (hunk 13 above)
Comment (remi-or) [reply src/transformers/generation/logits_process.py:140]: But any time a new logit processor is added, its support is set to `None` that way. A mapping we will have to manually keep updating which is less redundant. wdyt?

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 2 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: We do? I made this more clear.

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 7 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py:179]: The issue is not really the typing, it's more about users being aware of which kwarg are ignored + erroring out if a user wants to tune temperature when the temperature logit wrapper is not present. The issue with CB is that once tensors are created, we cannot go back, since they are static. So adding a logit processor on the fly would invalidate all cuda graphs. If we want to add this feature later on why not, but until then I think it's better to have rigorous type checks  and no silent fails.

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 8 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py:189]: Yes!

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 1 above)
Comment (ArthurZucker) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: I mean the kwargs you can list which ones are supporterd / processor is what I mean

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py:179]: Got it got it, the cost paid by the user would be too big agreed

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 9 above)
Comment (ArthurZucker) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: Typed kwargs is what I had in mind good lord@!

Comment (ArthurZucker) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: but the argument of cost of init is fair

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 10 above)
Comment (ArthurZucker) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: for dataclasses not config necessarily no?

**Review by ArthurZucker — APPROVED** [review]

File: src/transformers/generation/continuous_batching/cb_logits_processors.py (hunk 10 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cb_logits_processors.py]: But logits processors are not dataclasses? Anyway, we initialize it from the classic logit processor now, so an error in parameters will be caught by the classic logits processors before CB one is even initialized
